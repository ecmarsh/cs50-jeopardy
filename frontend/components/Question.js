import React, { Component, Fragment, useState } from 'react';
import PropTypes from 'prop-types';
import { max } from 'ramda';
import formatScore from '../lib/formatScore';
import Button from './styles/Button';
import { UpdateTeamScoreButton } from './ScoreTeams';
import { Divider } from './styles/Utilities';
import DailyDouble from './DailyDouble';
import { Grow, Shift, Unfocus } from './Animated';

class GamePlayQuestion extends Component {
  state = {
    chosenTeam: {},
    showAnswer: false,
    gen: this.stages(),
  };

  *stages() {
    const {
      gameName,
      round,
      category,
      categoryQuestion,
      toggleChosenTeam,
      toggleView,
      teams,
    } = this.props;

    // CHOOSE ANSWERING TEAM
    yield (
      <Fragment key={`stage1`}>
        <Shift cName="controls__message">Which team is answering?</Shift>
        <div className="controls__ctas">
          {teams.map((team, idx) => (
            <Button
              key={'teambutton-' + idx}
              data-test={'cta-button-' + idx}
              game
              onClick={e =>
                this.setChosenTeam(e, idx, team.id, team.name, team.score)
              }
            >
              Team {team.name}
            </Button>
          ))}
        </div>
      </Fragment>
    );
    // AWAITING DOUBLE JEOPARDY WAGER
    if (categoryQuestion.isDouble) {
      yield <Shift cName="controls__message">Waiting for wager...</Shift>;
    }
    // REVEAL ANSWER
    yield (
      <Fragment key={`stage2`}>
        <Shift cName="controls__message">Waiting for answer...</Shift>
        <Button
          data-test="cta-button-1"
          game
          onClick={() => {
            this.setState({ showAnswer: true });
          }}
        >
          Show Answer
        </Button>
      </Fragment>
    );
    // RIGHT OR WRONG
    yield (
      <Fragment key={`stage3`}>
        <Shift cName="controls__message">Were they correct?</Shift>
        <div className="controls__ctas">
          {[1, -1].map((multiplier, n) => (
            <UpdateTeamScoreButton
              key={'rightwrong' + n}
              data-test={'cta-button-' + n}
              dataTest={'cta-button-' + n}
              gameName={gameName}
              chosenTeam={this.state.chosenTeam}
              sumScoreWith={
                (this.state.override ||
                  categoryQuestion.difficulty * (round * 200)) * multiplier
              }
              categoryId={category.id}
              categoryQuestionId={categoryQuestion.id}
              toggleView={toggleView}
              toggleChosenTeam={toggleChosenTeam}
            >
              {multiplier === 1 ? 'Correct' : 'Incorrect'}
            </UpdateTeamScoreButton>
          ))}
        </div>
      </Fragment>
    );
    // Updating the score will unmount component
    return null;
  }

  setChosenTeam = (e, idx, id, name, score) => {
    e.preventDefault();
    const chosenTeam = { idx, id, name, score };
    this.props.toggleChosenTeam(id);
    this.setState(prevState => ({
      teamShown: !prevState.teamShown,
      chosenTeam,
    }));
  };

  overrideQuestionPrice = override => this.setState({ override }); // set 2x wager

  render() {
    const { round, category, categoryQuestion, active } = this.props;
    const questionPrice =
      this.state.override || categoryQuestion.difficulty * (round * 200);
    const { chosenTeam, gen, showAnswer } = this.state;

    return (
      <Grow active={active}>
        <div className="fullscreen__inner">
          <div className="titles">
            <div className="titles__inner">
              <h3>{category.name}</h3>
              <h3>{formatScore(questionPrice)}</h3>
            </div>
            <Divider />
          </div>

          <div className="main__display">
            {categoryQuestion.isDouble && !this.state.override && (
              <DailyDouble
                setQuestionPrice={this.overrideQuestionPrice}
                override={this.state.override}
                isChosenTeam={Boolean(chosenTeam.id)}
                minWager={round * 200}
                maxWager={max(chosenTeam.score, round * 1000)}
              />
            )}
            {(!categoryQuestion.isDouble || this.state.override) && (
              <>
                <Unfocus off={showAnswer}>
                  {categoryQuestion.question.question}
                </Unfocus>
                {showAnswer && (
                  <Shift cName="answer">
                    {categoryQuestion.question.answer}
                  </Shift>
                )}
              </>
            )}
          </div>

          <div className="controls">
            <Divider />
            {gen.next().value}
          </div>
        </div>
      </Grow>
    );
  }
}

export default GamePlayQuestion;
