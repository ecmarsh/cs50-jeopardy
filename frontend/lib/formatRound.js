export default function(page) {
  let title;
  switch (page) {
    case 1:
      title = 'Single';
      break;
    case 2:
      title = 'Double';
      break;
    case 3:
      title = 'Final';
      break;
    default:
      title = 'CS50';
  }

  return title + ' Jeopardy';
}
