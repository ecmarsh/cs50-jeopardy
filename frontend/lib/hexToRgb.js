export default function(hex) {
  const chunk = 2,
    radix = 16,
    rgb = [];

  let i = 1;
  while (hex && i < hex.length) {
    rgb.push(parseInt(hex.substring(i, i + chunk), radix));
    i += chunk;
  }

  return rgb.toString();
}
