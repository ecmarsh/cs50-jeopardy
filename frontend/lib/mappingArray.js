export default function(N = 0, isFilled = true, startAtZero = false) {
  if (N == 0) return [];
  if (N > 0 && !isFilled) return Array(N).fill();
  if (isFilled && !startAtZero)
    return Array.from({ length: N }, (v, k) => k + 1);
  return Array.from({ length: N }, (v, k) => k++);
}
