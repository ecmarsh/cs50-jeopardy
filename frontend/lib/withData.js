import withApollo from 'next-with-apollo';
import ApolloClient from 'apollo-boost';
import { endpoint, endpointProd } from '../config';
import { LOCAL_STATE_QUERY } from '../components/SoundEffect';
import { ANSWERED_QUESTIONS_QUERY } from '../components/GameScreen';
import gql from 'graphql-tag';

// Setup for SSR w/ next and config endpoint
function createClient({ headers }) {
  const typeDefs = gql`
    type StudyQuestion {
      data: Json
    }
  `;
  const client = new ApolloClient({
    uri: process.env.NODE_ENV === 'development' ? endpoint : endpointProd,
    request: operation => {
      operation.setContext({
        fetchOptions: {
          credentials: 'include',
        },
        headers,
      });
    },
    typeDefs,
    // cache
    clientState: {
      resolvers: {
        Mutation: {
          toggleSound(_, variables, { cache }) {
            // read the soundOn value from the cache
            const { soundOn } = cache.readQuery({
              query: LOCAL_STATE_QUERY,
            });
            // Toggle sound on || off
            const data = {
              data: { soundOn: !soundOn },
            };
            cache.writeData(data);
            return data;
          },
          addAnswered(_, { id, val }, { cache }) {
            // read the answered questions from the cache
            const { answeredQuestions } = cache.readQuery({
              query: ANSWERED_QUESTIONS_QUERY,
            });
            const newQuestion = { id, val, __typename: 'StudyQuestion' };
            const jsonified = JSON.stringify([
              newQuestion,
              ...answeredQuestions,
            ]);
            // Add new question to data
            const data = {
              data: {
                answeredQuestions: [newQuestion, ...answeredQuestions],
              },
            };

            cache.writeData(data);
            return data;
          },
          resetBoard(_, variables, { cache }) {
            // Reset answered questions
            const data = {
              data: {
                answeredQuestions: [],
              },
            };
            cache.writeData(data);
            return data;
          },
        },
      },
      defaults: {
        soundOn: false,
        answeredQuestions: [],
      },
    },
  });
  return client;
}

export default withApollo(createClient);
