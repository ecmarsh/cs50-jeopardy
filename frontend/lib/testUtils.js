import casual from 'casual';
import mappingArray from './mappingArray';

// seed it for repeatable random sequence
casual.seed(123);

const fakeQuestion = () => ({
  id: 'abc123',
  question: 'Time complexity of merge sort?',
  answer: 'n log n',
  user: {
    id: casual.uuid,
    name: 'User Name',
    __typename: 'User',
  },
  createdAt: '2019-02 - 26T19: 24: 16.000Z',
  game: {
    id: 'game123',
    name: 'GameForGeekz',
    user: {
      id: casual.uuid,
      __typename: 'User',
    },
    isPublic: casual.coin_flip,
    __typename: 'Game',
  },
  categoryQuestions: [
    {
      id: 'catQuest123',
      category: { id: 'cat123', __typename: 'Category' },
      __typename: '__CategoryQuestion',
    },
  ],
  __typename: 'Question',
});

const fakeUser = () => ({
  __typename: 'User',
  id: 'userABC123',
  name: casual.name,
  email: casual.email,
  permissions: ['USER'],
});

const fakeAdmin = () => ({
  __typename: 'User',
  id: 'adminABC123',
  name: casual.name,
  email: casual.email,
  permission: ['ADMIN', 'USER'],
  games: [],
});

const fakeCategoryQuestion = (id = 'catQuest123', difficulty = 1) => ({
  id,
  difficulty,
  isDouble: false,
  answered: casual.coin_flip,
  question: fakeQuestion(),
  category: { id: 'cat123', name: casual.word, __typename: 'Category' },
  game: { id: 'g@m3', name: 'Test Game', __typename: 'Game' },
  __typename: 'CategoryQuestion',
});

const fillCategoryQuestions = (n = 5) =>
  mappingArray(n).map(i => fakeCategoryQuestion(`catQuest-${i}`, i));

const fakeCategory = (id = 'cat123') => ({
  id,
  name: casual.word,
  game: { id: 'g@m3', name: 'Test Game' },
  categoryQuestions: fillCategoryQuestions(),
  __typename: 'Category',
});

const fillCategories = (n = 11) =>
  mappingArray(n).map(i => fakeCategory(`cat-${i}`));

const fakeTeam = id => ({
  __typename: 'Team',
  id: `t${id}`,
  name: `Team${id}`,
  score: 1000,
  game: { id: 'g@m3', name: 'Test Game' },
});

const fakeGameConfig = () => ({
  id: 'config123',
  hasDoubleJeopardy: false,
  hasRoundTimer: false,
  roundTime: 0,
  finalTime: 0,
  __typename: 'GameConfig',
});

const fakeGame = () => ({
  __typename: 'Game',
  id: 'g@m3',
  name: 'Test Game',
  createdAt: '2019-03 - 06T19: 12: 16.000Z',
  isPublic: casual.coin_flip,
  user: fakeUser(),
  teams: [fakeTeam(1), fakeTeam(2)],
  categories: fillCategories(),
  questions: [fakeQuestion()],
  config: fakeGameConfig(),
});

const fakeStudyGame = (id = 'SG123') => ({
  __typename: 'StudyGame',
  id,
  answeredQuestions: [fakeCategoryQuestion(), fakeCategoryQuestion()],
  score: 1000,
  game: fakeGame(),
  user: fakeUser(),
  createdAt: '2019-04 - 25T19: 10: 16.000Z',
});

const fakeGameSummary = () => ({
  questionsConnection: {
    aggregate: {
      count: casual.double + 1,
      __typename: 'aggregate',
    },
    __typename: 'questionsConnection',
  },
  categoryQuestionsConnection: {
    aggregate: {
      count: casual.double,
      __typename: 'aggregate',
    },
    __typename: 'categoryQuestionsConnection',
  },
});

// Fake LocalStorage
class LocalStorageMock {
  constructor() {
    this.store = {};
  }

  clear() {
    this.store = {};
  }

  getItem(key) {
    return this.store[key] || null;
  }

  setItem(key, value) {
    this.store[key] = value.toString();
  }

  removeItem(key) {
    delete this.store[key];
  }
}

export {
  LocalStorageMock,
  fakeQuestion,
  fakeUser,
  fakeAdmin,
  fakeCategory,
  fakeCategoryQuestion,
  fakeTeam,
  fakeGame,
  fakeStudyGame,
  fakeGameSummary,
  fakeGameConfig,
};
