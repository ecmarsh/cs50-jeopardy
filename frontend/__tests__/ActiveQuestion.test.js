import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import wait from 'waait';
import Question from '../components/Question';
import { fakeGame, fakeCategory } from '../lib/testUtils';

const game = fakeGame();
const category = updatedFakeCategory();

const mockProps = {
  gameName: game.name,
  round: 1,
  category: category.id,
  categoryQuestion: category.categoryQuestions[0],
  toggleChosenTeam: jest.fn(),
  toggleView: jest.fn(),
  teams: game.teams,
  active: true,
};

describe('<GameQuestion />', () => {
  beforeEach(() => {
    mockProps.toggleView.mockClear();
    mockProps.toggleChosenTeam.mockClear();
  });
  it('renders active question and matches snapshot', async () => {
    const wrapper = shallow(<Question {...mockProps} />);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
  it('renders next stage after choosing team', async () => {
    const wrapper = shallow(<Question {...mockProps} />);
    const button = wrapper.find('Button[data-test="cta-button-1"]');
    expect(
      Object.getOwnPropertyNames(wrapper.instance().state.chosenTeam)
    ).toHaveLength(0);
    button.simulate('click', {
      preventDefault: jest.fn(),
    });
    await wait();
    wrapper.update();

    expect(
      Object.getOwnPropertyNames(wrapper.instance().state.chosenTeam)
    ).toHaveLength(4);
    expect(wrapper.find('Button[data-test="cta-button-1"]').text()).toContain(
      'Show Answer'
    );
  });
  it('renders answer when show answer is clicked', async () => {
    const wrapper = shallow(<Question {...mockProps} />);
    const btn = 'Button[data-test="cta-button-1"]';

    wrapper.find(btn).simulate('click', {
      preventDefault: jest.fn(),
    });
    await wait();
    wrapper.update();
    const answer = wrapper
      .find('Shift')
      .filterWhere(n => n.get(0).props.cName === 'answer');
    expect(answer).toHaveLength(0);
    wrapper.find(btn).simulate('click', {
      preventDefault: jest.fn(),
    });
    await wait();
    wrapper.update();
    const answer2 = wrapper
      .find('Shift')
      .filterWhere(n => n.get(0).props.cName === 'answer');
    expect(answer2).toHaveLength(1);
    expect(answer2.children().text()).toContain(
      mockProps.categoryQuestion.question.answer
    );

    expect(
      wrapper
        .find(`UpdateTeamScore${btn}`)
        .children()
        .text()
    ).toContain('Incorrect');
  });
});

function updatedFakeCategory() {
  const _categories = fakeCategory();
  // Ensure both questions are available
  _categories.categoryQuestions[0].answered = false;
  _categories.categoryQuestions[1].answered = false;
  // Set on/off daily double setting
  _categories.categoryQuestions[0].isDouble = false;
  _categories.categoryQuestions[1].isDouble = true;

  return _categories;
}
