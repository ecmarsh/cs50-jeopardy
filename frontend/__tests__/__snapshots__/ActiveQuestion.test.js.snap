// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GameQuestion /> renders active question and matches snapshot 1`] = `
<Component
  active={true}
>
  <div
    className="fullscreen__inner"
  >
    <div
      className="titles"
    >
      <div
        className="titles__inner"
      >
        <h3 />
        <h3>
          $200
        </h3>
      </div>
      <Utilities__Divider />
    </div>
    <div
      className="main__display"
    >
      <Unfocus
        off={false}
      >
        Time complexity of merge sort?
      </Unfocus>
    </div>
    <div
      className="controls"
    >
      <Utilities__Divider />
      <Shift
        cName="controls__message"
      >
        Which team is answering?
      </Shift>
      <div
        className="controls__ctas"
      >
        <Button
          data-test="cta-button-0"
          game={true}
          key="teambutton-0"
          onClick={[Function]}
        >
          Team 
          Team1
        </Button>
        <Button
          data-test="cta-button-1"
          game={true}
          key="teambutton-1"
          onClick={[Function]}
        >
          Team 
          Team2
        </Button>
      </div>
    </div>
  </div>
</Component>
`;
